fs/socket/socket.c

```diff
@@ -0,0 +1,298 @@
+  if (sockfd < 0)
+    {
+      kmm_free(*psock);
+    }
+
+  inode_addref(&g_sock_inode);
+
+  return sockfd;
+}
+
+/****************************************************************************
+ * Name: sockfd_socket
+ *
+ * Description:
+ *   Given a socket descriptor, return the underlying socket structure.
+ *
+ * Input Parameters:
+ *   sockfd - The socket descriptor index to use.
+ *
+ * Returned Value:
+ *   On success, a reference to the socket structure associated with the
+ *   the socket descriptor is returned.  NULL is returned on any failure.
+ *
+ ****************************************************************************/
+
+FAR struct socket *sockfd_socket(int sockfd)
+{
+  FAR struct file *filep;
+
+  if (fs_getfilep(sockfd, &filep) < 0)
+    {
+      return NULL;
+    }
+
+  if (INODE_IS_SOCKET(filep->f_inode))
+    {
+      return filep->f_priv;
+    }
+
+  return NULL;
+}
+
+/****************************************************************************
+ * Name: socket
+ *
+ * Description:
+ *   socket() creates an endpoint for communication and returns a descriptor.
+ *
+ * Input Parameters:
+ *   domain   (see sys/socket.h)
+ *   type     (see sys/socket.h)
+ *   protocol (see sys/socket.h)
+ *
+ * Returned Value:
+ *   A non-negative socket descriptor on success; -1 on error with errno set
+ *   appropriately.
+ *
+ *   EACCES
+ *     Permission to create a socket of the specified type and/or protocol
+ *     is denied.
+ *   EAFNOSUPPORT
+ *     The implementation does not support the specified address family.
+ *   EINVAL
+ *     Unknown protocol, or protocol family not available.
+ *   EMFILE
+ *     Process file table overflow.
+ *   ENFILE
+ *     The system limit on the total number of open files has been reached.
+ *   ENOBUFS or ENOMEM
+ *     Insufficient memory is available. The socket cannot be created until
+ *     sufficient resources are freed.
+ *   EPROTONOSUPPORT
+ *     The protocol type or the specified protocol is not supported within
+ *     this domain.
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+int socket(int domain, int type, int protocol)
+{
+  FAR struct socket *psock;
+  FAR struct file *filep;
+  int errcode;
+  int sockfd;
+  int ret;
+
+  /* Allocate a socket descriptor */
+
+  sockfd = sockfd_allocate(&psock);
+  if (sockfd < 0)
+    {
+      nerr("ERROR: Failed to allocate a socket descriptor\n");
+      errcode = -sockfd;
+      goto errout;
+    }
+
+  ret = fs_getfilep(sockfd, &filep);
+  if (ret < 0)
+    {
+      goto errout_with_sockfd;
+    }
+
+  if (type & SOCK_CLOEXEC)
+    {
+      filep->f_oflags |= O_CLOEXEC;
+    }
+
+  /* Initialize the socket structure */
+
+  ret = psock_socket(domain, type, protocol, psock);
+  if (ret < 0)
+    {
+      nerr("ERROR: psock_socket() failed: %d\n", ret);
+      errcode = -ret;
+      goto errout_with_sockfd;
+    }
+
+  return sockfd;
+
+errout_with_sockfd:
+  close(sockfd);
```

Comment:
We should not call the close(), since it sets a cancellation point.
```suggestion
  nx_close(sockfd);
```

Reply:
Done!